// defuddle/dist/extractors/nytimes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NytimesExtractor = void 0;
const _base_1 = require("./_base");
const dom_1 = require("../utils/dom");
const INJECTED_ATTR = 'data-defuddle-nyt';
class NytimesExtractor extends _base_1.BaseExtractor {
    constructor(document, url, schemaOrgData, options) {
        super(document, url, schemaOrgData, options);
        this.preloadedData = null;
        this.contentSelector = null;
        this.preloadedData = this.extractPreloadData();
        if (this.preloadedData) {
            const body = this.preloadedData.sprinkledBody || this.preloadedData.body;
            if (body?.content?.length) {
                // Inject rendered content into the document for the pipeline
                const existing = document.querySelector(`[${INJECTED_ATTR}]`);
                if (!existing) {
                    const wrapper = document.createElement('div');
                    wrapper.setAttribute(INJECTED_ATTR, '');
                    wrapper.appendChild((0, dom_1.parseHTML)(document, this.renderBlocks(body.content)));
                    document.body.appendChild(wrapper);
                }
                this.contentSelector = `[${INJECTED_ATTR}]`;
            }
        }
    }
    canExtract() {
        return this.contentSelector !== null;
    }
    extract() {
        const article = this.preloadedData;
        const title = article.headline?.default || '';
        const authors = (article.bylines?.[0]?.creators || [])
            .map(c => c.displayName)
            .filter(Boolean)
            .join(', ');
        const published = article.firstPublished || '';
        const description = article.summary || '';
        return {
            content: '',
            contentHtml: '',
            contentSelector: this.contentSelector,
            variables: {
                title,
                author: authors,
                published,
                description,
            },
        };
    }
    extractPreloadData() {
        const scripts = this.document.querySelectorAll('script:not([src])');
        for (const script of scripts) {
            const text = script.textContent || '';
            if (!text.includes('window.__preloadedData'))
                continue;
            const match = text.match(/window\.__preloadedData\s*=\s*({[\s\S]+})\s*;?\s*$/);
            if (!match)
                continue;
            try {
                // NYT JSON contains JS `undefined` values — replace with null
                const raw = match[1].replace(/(?<=:)undefined(?=[,}\]])/g, 'null');
                const data = JSON.parse(raw);
                return data.initialData?.data?.article || null;
            }
            catch {
                return null;
            }
        }
        return null;
    }
    renderBlocks(blocks) {
        const parts = [];
        for (const block of blocks) {
            switch (block.__typename) {
                case 'ParagraphBlock':
                    parts.push(`<p>${this.renderInlines(block.content)}</p>`);
                    break;
                case 'Heading2Block':
                    parts.push(`<h2>${this.renderInlines(block.content)}</h2>`);
                    break;
                case 'Heading3Block':
                    parts.push(`<h3>${this.renderInlines(block.content)}</h3>`);
                    break;
                case 'Heading4Block':
                    parts.push(`<h4>${this.renderInlines(block.content)}</h4>`);
                    break;
                case 'ImageBlock': {
                    const img = block;
                    const media = img.media;
                    if (!media)
                        break;
                    const src = this.getBestImageUrl(media);
                    if (!src)
                        break;
                    const alt = this.escapeAttr(media.altText || media.caption?.text || '');
                    const caption = media.caption?.text || '';
                    const credit = media.credit || '';
                    const figcaptionParts = [caption, credit].filter(Boolean);
                    if (figcaptionParts.length) {
                        parts.push(`<figure>` +
                            `<img src="${this.escapeAttr(src)}" alt="${alt}">` +
                            `<figcaption>${this.escapeHtml(figcaptionParts.join(' '))}</figcaption>` +
                            `</figure>`);
                    }
                    else {
                        parts.push(`<img src="${this.escapeAttr(src)}" alt="${alt}">`);
                    }
                    break;
                }
                case 'HeaderBasicBlock':
                case 'Dropzone':
                    // Skip structural blocks
                    break;
                default: {
                    // Handle unknown blocks that have inline content
                    const generic = block;
                    if (generic.content?.length) {
                        parts.push(`<p>${this.renderInlines(generic.content)}</p>`);
                    }
                    break;
                }
            }
        }
        return parts.join('\n');
    }
    renderInlines(inlines) {
        if (!inlines)
            return '';
        return inlines.map(inline => {
            let text = this.escapeHtml(inline.text || '');
            if (!inline.formats?.length)
                return text;
            for (const fmt of inline.formats) {
                switch (fmt.__typename) {
                    case 'BoldFormat':
                        text = `<strong>${text}</strong>`;
                        break;
                    case 'ItalicFormat':
                        text = `<em>${text}</em>`;
                        break;
                    case 'LinkFormat':
                        if (fmt.url) {
                            text = `<a href="${this.escapeAttr(fmt.url)}">${text}</a>`;
                        }
                        break;
                }
            }
            return text;
        }).join('');
    }
    getBestImageUrl(media) {
        const crops = media?.crops;
        if (!crops?.length)
            return null;
        // Prefer superJumbo > jumbo > articleLarge
        const preferred = ['superJumbo', 'jumbo', 'articleLarge'];
        for (const name of preferred) {
            for (const crop of crops) {
                const rendition = crop.renditions?.find((r) => r.name === name);
                if (rendition?.url)
                    return rendition.url;
            }
        }
        // Fall back to any rendition with a URL
        for (const crop of crops) {
            if (crop.renditions?.length && crop.renditions[0].url) {
                return crop.renditions[0].url;
            }
        }
        return null;
    }
    escapeHtml(text) {
        return text
            .replace(/&/g, '&amp;')
            .replace(/</g, '&lt;')
            .replace(/>/g, '&gt;');
    }
    escapeAttr(text) {
        return text
            .replace(/&/g, '&amp;')
            .replace(/"/g, '&quot;')
            .replace(/</g, '&lt;')
            .replace(/>/g, '&gt;');
    }
}
exports.NytimesExtractor = NytimesExtractor;
